gitclone/steperror.go

```diff
@@ -1,11 +1,178 @@
+	m := PatternErrorMatcher{
+		handlers:       handlers,
+		defaultHandler: defaultHandler,
+	}
+
+	return &m
+}
+
+// Run ...
+func (m *PatternErrorMatcher) Run(msg string) step.Recommendation {
+	for pattern, handler := range m.handlers {
+		re := regexp.MustCompile(pattern)
+		if re.MatchString(msg) {
+			matches := re.FindStringSubmatch((msg))
+
+			if len(matches) > 1 {
+				matches = matches[1:]
+			}
+
+			if matches != nil {
+				detail := handler(matches...)
+				return newDetailedErrorRecommendation(detail)
+			}
+		}
+	}
+
+	detail := m.defaultHandler(msg)
+	return newDetailedErrorRecommendation(detail)
+}
+
+func newCheckoutFailedPatternErrorMatcher() *PatternErrorMatcher {
+	return newPatternErrorMatcher(
+		checkoutFailedGenericDetail,
+		map[string]DetailBuilder{},
+	)
+}
+
+func newFetchFailedPatternErrorMatcher() *PatternErrorMatcher {
+	return newPatternErrorMatcher(
+		fetchFailedGenericDetail,
+		map[string]DetailBuilder{
+			`Permission denied (.+)\.`:                                                               fetchFailedSSHAccessError,
+			`fatal: repository '(.+)' not found`:                                                     fetchFailedCouldNotFindGitRepository,
+			`fatal: '(.+)' does not appear to be a git repository`:                                   fetchFailedCouldNotFindGitRepository,
+			`fatal: (.+)/info/refs not valid: is this a git repository?`:                             fetchFailedCouldNotFindGitRepository,
+			`remote: HTTP Basic: Access denied[\n]*fatal: Authentication failed for '(.+)'`:          fetchFailedHTTPAccessError,
+			`remote: Invalid username or password\(\.\)[\n]*fatal: Authentication failed for '(.+)'`: fetchFailedHTTPAccessError,
+			`Unauthorized`:                          fetchFailedHTTPAccessError,
+			`Forbidden`:                             fetchFailedHTTPAccessError,
+			`remote: Unauthorized LoginAndPassword`: fetchFailedHTTPAccessError,
+			// `fatal: unable to access '(.+)': Failed to connect to .+ port \d+: Connection timed out
+			// `fatal: unable to access '(.+)': The requested URL returned error: 400`
+			// `fatal: unable to access '(.+)': The requested URL returned error: 403`
+			`fatal: unable to access '(.+)': (Failed|The requested URL returned error: \d+)`: fetchFailedHTTPAccessError,
+			// `ssh: connect to host (.+) port \d+: Connection timed out`
+			// `ssh: connect to host (.+) port \d+: Connection refused`
+			// `ssh: connect to host (.+) port \d+: Network is unreachable`
+			`ssh: connect to host (.+) port \d+:`:                                fetchFailedCouldConnectError,
+			`ssh: Could not resolve hostname (.+): Name or service not known`:    fetchFailedCouldConnectError,
+			`fatal: unable to access '.+': Could not resolve host: (\S+)`:        fetchFailedCouldConnectError,
+			`ERROR: The \x60(.+)' organization has enabled or enforced SAML SSO`: fetchFailedSamlSSOError,
+		})
+}
+
+func mapRecommendation(tag string, err error) step.Recommendation {
```

Comment:
as we only need the error message, I suggest passing a string only as input (this will help with the CI issue too)